main/src/hadd.cxx

```diff
@@ -90,284 +102,473 @@
-               }
-            }
-         }
-         ++ffirst;
-      } else if ( argv[a][0] == '-' ) {
-         bool farg = false;
-         if (force && argv[a][1] == 'f') {
-            // Bad argument
-            std::cerr << "Error: Using option " << argv[a] << " more than once is not supported.\n";
-            ++ffirst;
-            farg = true;
-         }
-         const char *prefix = "";
-         if (argv[a][1] == 'f' && argv[a][2] == 'k') {
-            farg = true;
-            force = kTRUE;
-            keepCompressionAsIs = kTRUE;
-            prefix = "k";
-         }
-         if (argv[a][1] == 'f' && argv[a][2] == 'f') {
-            farg = true;
-            force = kTRUE;
-            useFirstInputCompression = kTRUE;
-            if (argv[a][3] != '\0') {
-               std::cerr << "Error: option -ff should not have any suffix: " << argv[a] << " (suffix has been ignored)\n";
-            }
-         }
-         char ft[7];
-         for (int alg = 0; !useFirstInputCompression && alg <= 5; ++alg) {
-            for( int j=0; j<=9; ++j ) {
-               const int comp = (alg*100)+j;
-               snprintf(ft,7,"-f%s%d",prefix,comp);
-               if (!strcmp(argv[a],ft)) {
-                  farg = true;
-                  force = kTRUE;
-                  newcomp = comp;
-                  break;
-               }
-            }
-         }
-         if (!farg) {
-            // Bad argument
-            std::cerr << "Error: option " << argv[a] << " is not a supported option.\n";
+            Err() << "seen a positional argument '" << argRaw << "' after some flags."
+                     " Positional arguments were already parsed at this point (from '" << argv[args.fOutputArgIdx]
+                     << "' onwards), so you cannot pass more."
+                     " Please regroup your positional arguments so that hadd works as you expect.\n"
+                     "Cmdline: ";
+            for (int i = 0; i < argc; ++i)
+               std::cerr << argv[i] << " ";
+            std::cerr << "\n";
+
+            return {};
          }
-         ++ffirst;
-      } else if (!outputPlace) {
-         outputPlace = a;
       }
    }
 
+   return args;
+}
+
+int main(int argc, char **argv)
+{
+   if (argc < 3 || "-h" == std::string(argv[1]) || "--help" == std::string(argv[1])) {
+      fprintf(stderr, kCommandLineOptionsHelp);
+      return (argc == 2 && ("-h" == std::string(argv[1]) || "--help" == std::string(argv[1]))) ? 0 : 1;
+   }
+
+   const auto argsOpt = ParseArgs(argc, argv);
+   if (!argsOpt)
+      return 1;
+   const HAddArgs &args = *argsOpt;
+
+   ROOT::TIOFeatures features = args.fFeatures.value_or(ROOT::TIOFeatures{});
+   Int_t maxopenedfiles = args.fMaxOpenedFiles.value_or(0);
+   Int_t verbosity = args.fVerbosity.value_or(99);
+   TString cacheSize = args.fCacheSize.value_or("");
```

Comment:
As a visual aid to future readers, I would group here all the optional arguments and put a comment at the beginning of the section. I would also include in this section the `newcomp` variable probably